### ISSUE TITLE
Add polars parquet dataset abstraction

### ISSUE THREAD

**Issue by ion-elgreco** [pola-rs/polars#13086]
### Description

Currently it's not possible to read a parquet dataset that has mixed schema's per file but with one-top level schema, additionally we cannot provide some statistics regarding this dataset or each file. If Polars can create a similar abstraction to pyarrow dataset we could get even better support for delta tables that would allow better pushdown and read performance.

A valid delta table could contain, this schema in the first partition:

`OrderedDict([('foo', Int64), ('bar', Utf8)])`

and in the second partition:

`OrderedDict([('bar', Utf8), ('baz', Utf8')])`

The top level schema of the dataset is then `['foo', 'bar', 'baz'] `. 

Additionally we can provide statistics the each partition, on the first partition I could provide these statistics on each partition grabbed from the delta transaction log, which could be used by the polars optimizer to do pushdowns: 

file: '0-69b38df8-0c32-46b4-a67b-1a7d583eb9e3-0.parquet', `(((((is_valid(foo) and is_valid(bar)) and (foo >= 1)) and (bar >= "1")))`

**Comment by deanm0000** [comment]
It's your first partition joined to the second on your common columns or are they concatted with null values?

**Comment by ion-elgreco** [comment]
@deanm0000 they are concatenated, so in this case both are write actions to the delta table with `save_mode='append'`

**Comment by ion-elgreco** [comment]
@stinodego do you reckon something like this can be added? It could be a starting point to create a deltalake reader natively since the polars parquet reader is much faster and has better pushdown support than the polars->pyarrow scan

**Comment by coastalwhite** [comment]
Personally, I would like to see something like this. Maybe, this would be better added however as user known statistics per file. For example, you would do `pl.scan_parquet(multiple_files, dataset_statistics=df_with_statistics_per_file)`. This statistics dataframe would then have (possibly missing) `min`, `max`, `null_count` and `len` per file for the columns that it has knowledge for.

This would be easy to integrate into the existing way polars does stuff, and I think would provide most of the benefits of a dedicated `Dataset` class. Do you think this would help delta support @ion-elgreco?

**Comment by ion-elgreco** [comment]
@coastalwhite that can work, but we need to be clear on how these values are expected. A DeltaTable can return an arrow table of the add actions. Take this sample table:

```python
import polars as pl
from deltalake import DeltaTable

pl.DataFrame({"foo": [1,2,3]}).write_delta("data/test_table")
```
It would return this add action schema:

```python
path: string not null
size_bytes: int64 not null
modification_time: timestamp[ms] not null
data_change: bool not null
num_records: int64 not null
null_count: struct<foo: int64> not null
  child 0, foo: int64
min: struct<foo: int64> not null
  child 0, foo: int64
max: struct<foo: int64> not null
  child 0, foo: int64
```

Would that work for you? :)

**Comment by ion-elgreco** [comment]
@coastalwhite important to mention, in delta protocol you can tell which columns to explicitly use for collecting stats. So in the case that someone does `pl.col("baz") == 5` and it's not part of the stats hints, it means this column was not included in the stats collection, so you would need to filter based on the parquet stats for that column if it's not too expensive or just filter after reading all rowgroups for that col.

**Comment by coastalwhite** [comment]
Yes, I think that would work. This aligns quite closely with how we filter internally. I will soon take a look at whether we can add such an argument. It might be hidden for now, but delta should be able to use it.

**Comment by dkapitan** [comment]
+1 for the general idea to create a stats table.

**Comment by deanm0000** [comment]
If you're going to experiment, after 

```python
file_path = "data/test_table"
pl.DataFrame({"foo": [1,2,3]}).write_delta(file_path )
```

to get the stats table, you'd do

```python
delta_table = DeltaTable(file_path )
add_actions = delta_table.get_add_actions()
```

the path in that table will be relative to `file_path`.
